modules/core/DecodeError.scala

```diff
@@ -0,0 +1,174 @@
+      * `error` and repeatedly calling this method on the `error` if
+      * needed
+      */
+    @tailrec def deepError: DecodeError =
+      error match {
+        case error0: AtPath => error0.deepError
+        case _              => error
+      }
+  }
+
+  /** Qualifies a nested error indicating that it occurred while decoding
+    * a particular index within the source traversable data structure
+    */
+  final case class AtIndex(index: Int, error: DecodeError) extends DecodeError
+
+  /** Indicates that an underlying (and generally unexpected) error occurred
+    * while interacting with the source data structure.
+    *
+    * This typically shouldn't happen and this error may be removed in
+    * future releases.
+    */
+  final case class Underlying(underlying: Throwable) extends LeafDecodeError {
+    def canEqual(a: Any) = a.isInstanceOf[Underlying]
+    override def equals(that: Any): Boolean =
+      that match {
+        case that: Underlying => that.canEqual(this) && this.hashCode == that.hashCode
+        case _                => false
+      }
+    override def hashCode: Int =
+      Option(underlying.getMessage).hashCode
+  }
+
+  /** An aggregation of multiple child errors that ocurred
+    */
+  sealed abstract class AggregateError extends DecodeError {
+
+    /** The first child error that occurred */
+    def head: DecodeError
+
+    /** All of the other child errors that occurred */
+    def tail: List[DecodeError]
+
+    final override def isAggregate = true
+
+    override def toString: String =
+      s"""${getClass.getSimpleName}(${(head :: tail).mkString(", ")})"""
+  }
+
+  /** A conjunction of child errors.
+    *
+    * All child errors must be resolved for decoding to succeed.
+    */
+  final case class And(head: DecodeError, tail: List[DecodeError]) extends AggregateError
+  object And {
+    def apply(head: DecodeError, tail: DecodeError*): And = And(head, tail.toList)
+  }
+
+  /** A disjunction of child errors.
+    *
+    * One child error must be resolved for decoding to succeed.
+    */
+  final case class Or(head: DecodeError, tail: List[DecodeError]) extends AggregateError
+  object Or {
+    def apply(head: DecodeError, tail: DecodeError*): Or = Or(head, tail.toList)
+  }
+
+  /** An identity value that does nothing when `and` or `or`'d with
+    * another error
+    */
+  private[classy] case object Identity extends DecodeError
+
+  /** Create a conjunction of errors.
+    *
+    * If either error is already a conjunction then the contents are
+    * unfolded before returning a single "flattened" result.
+    */
+  def and(a: DecodeError, b: DecodeError): DecodeError = (a, b) match {
+    case (anyA, Identity) => anyA
+    case (Identity, anyB) => anyB
+    case (manyA: And, manyB: And) =>
+      And(manyA.head, manyA.tail ::: manyB.head :: manyB.tail)
+    case (manyA: And, oneB) =>
+      And(manyA.head, manyA.tail ::: oneB :: Nil)
+    case (oneA, manyB: And) =>
```

Comment:
I think we should pattern match for `And(headA, tailA)` (on the constructor) instead of the type (`many: And`), similar for `Or` below.